### ISSUE TITLE
Operations like Or fail race checks

### ISSUE THREAD

**Issue by cvermilion** [RoaringBitmap/roaring#78]
In practice this may not be dangerous, but concurrently performing operations like `Or(rb1, rb2)` triggers warnings in Go's race detector due to concurrently writing the `roaringarray.needCopyOnWrite` flags on the argument bitmaps.

Example test:

```go
func TestConcurrentBitmapAccess(t *testing.T) {
	rb1 := roaring.BitmapOf(0, 1, 10)

	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			rb2 := roaring.NewBitmap()
			rb2.Or(rb1)
			wg.Done()
		}()
	}
	wg.Wait()
}
```

Run via `go test -race` and you should see race warnings like:

```
==================
WARNING: DATA RACE
Write at 0x00c4201436e8 by goroutine 9:
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.(*roaringArray).appendCopyMany()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaringarray.go:91 +0x3f6
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.Or()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaring.go:686 +0xa73
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.(*Bitmap).Or()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaring.go:578 +0x46
  github.com/figly/datasearch/figsearch.TestConcurrentBitmapAccess.func1()
      /Users/verm/go/src/github.com/figly/datasearch/figsearch/codec_test.go:67 +0x2e5

Previous write at 0x00c4201436e8 by goroutine 8:
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.(*roaringArray).appendCopyMany()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaringarray.go:91 +0x3f6
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.Or()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaring.go:686 +0xa73
  github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring.(*Bitmap).Or()
      /Users/verm/go/src/github.com/figly/datasearch/vendor/github.com/RoaringBitmap/roaring/roaring.go:578 +0x46
  github.com/figly/datasearch/figsearch.TestConcurrentBitmapAccess.func1()
      /Users/verm/go/src/github.com/figly/datasearch/figsearch/codec_test.go:67 +0x2e5

Goroutine 9 (running) created at:
  github.com/figly/datasearch/figsearch.TestConcurrentBitmapAccess()
      /Users/verm/go/src/github.com/figly/datasearch/figsearch/codec_test.go:69 +0x115
  testing.tRunner()
      /usr/local/Cellar/go/1.7.1/libexec/src/testing/testing.go:610 +0xc9

Goroutine 8 (finished) created at:
  github.com/figly/datasearch/figsearch.TestConcurrentBitmapAccess()
      /Users/verm/go/src/github.com/figly/datasearch/figsearch/codec_test.go:69 +0x115
  testing.tRunner()
      /usr/local/Cellar/go/1.7.1/libexec/src/testing/testing.go:610 +0xc9
==================
```

A workaround is to call `rb.Clone()` before using in any other operations in a separate goroutine.

Possible solutions:

1. Leave as is but clarify this behavior in documentation. The current documentation gives me the impression that reading bitmaps in concurrent goroutines should be safe if I haven't set them to be copy-on-write. The race warning suggests otherwise. This may be wrong (the racing goroutines are setting the memory to the same value) but at minimum it's worrying and some doc mention would be helpful.

1. Use atomic loads/stores when accessing that flag. At minimum using an atomic store to implement `roaringarray.setNeedsCopyOnWrite` fixes this particular race for me; I'm not sure if other accesses of the flag need similar treatment.

1. Have operations like `Or` use `appendWithoutCopy` and `appendWithoutCopyMany` if the argument bitmaps aren't copy-on-write. (I haven't thought about what the performance indications are here.)

**Comment by lemire** [comment]
Excellent. This issue should get resolved.

It is, somewhat, already documented in our README...

https://github.com/RoaringBitmap/roaring#thread-safety

But we can do better. Right now, @glycerine is working on some major changes, so I would rather that we delay changing the code.

**Comment by cvermilion** [comment]
Sounds good! No rush on our end; as I noted the workaround is pretty simple and performance isn't a big issue yet. But I had to figure it out because we've added race checking to our CI pipeline...

**Comment by lemire** [comment]
@cvermilion  Sure. I think it should be possible to have the best of all words... no data loss, no data race and no performance penalty whatsoever. All of it without having to produce a lot of code.

**Comment by glycerine** [comment]
Bitmap.Or() is defined as updating its receiver in place:

// Or computes the union between two bitmaps and stores the result in the current bitmap

so I would not expect Or() to be safe to access across multiple goroutines at all.

**Comment by cvermilion** [comment]
Right, but it's not the receiver that's the problem -- `rb2.Or(rb1)` modifies `rb1`, too (as I understand it above).

**Comment by glycerine** [comment]
The receiver will be another problem, even if the race detector doesn't catch it (it can miss things if they aren't tested). The right thing here is to clarify the expectations.  Nobody should expect Or() to be multi-goroutine safe.

**Comment by lemire** [comment]
Maybe we should focus on ``roaring.Or`` as a starting point.

https://godoc.org/github.com/RoaringBitmap/roaring#Or

**Comment by cvermilion** [comment]
Right; I would expect `roaring.Or(rb1, rb2)` to be safe to call in multiple goroutines at once but my understanding is that currently it isn't.

Meanwhile I definitely wouldn't expect `rb1.Or(rb2)` to be safe when called with the same `rb1` in multiple goroutines, but I _would_ naively expect it to be safe when called with distinct `rb1` values but a shared `rb2`.

**Comment by lemire** [comment]
Now that @glycerine's massive PR has been merged (creating version 0.3.0), we could safely work on this.

**Comment by lemire** [comment]
@cvermilion 

Can you check out this commit...
https://github.com/RoaringBitmap/roaring/commit/d510ddf8795919278c281125ddb97e614ba15f92

It should do what you want... It certainly passes your test. 

I'll wait before issuing a release to see if it is a progress from your point of view.

**Comment by cvermilion** [comment]
@lemire, it fixes the races I'd seen in my code, awesome! I'm not familiar enough with the internals of the library to judge whether it solves the general case, but it looks good to me.

**Comment by lemire** [comment]
Ok, released in 0.3.2.
